calico_node/startup/startup.go

```diff
@@ -906,6 +913,84 @@ func ensureDefaultConfig(ctx context.Context, cfg *apiconfig.CalicoAPIConfig, c
 	return nil
 }
 
+type ClientV1 string
+
+func (_ ClientV1) Blah() (string, bool, error) {
+	return "v2.6.1", true, nil
+}
+
+// ensureMigrated ensures any data migration needed is done.
+func ensureMigrated(ctx context.Context, cfg *apiconfig.CalicoAPIConfig, c client.Interface) error {
+	var cv1 ClientV1
+
+	// Only run this migration if using KDD
+	if cfg.Spec.DatastoreType != apiconfig.Kubernetes {
+		switch v, ver := getClusterVersion(ctx, c, cv1); v {
+		case VUnknown:
+			// Unknown if the datastore has not been initialized (or problem accessing it)
+			// TODO: I know we don't want to make extra connections to the datastore but
+			// is there some connection test we should do here? -Erik
+			return nil
+		case V2PreV264:
+			return fmt.Errorf("Unable to migrate version %s to v3", ver)
+		case V2PostV264:
+			// Do migration
+			return nil
+		case V3orGreater:
+			// Already at the right version, nothing to do
+			return nil
+		}
+	}
+	return nil
+}
+
+type VerClass int
+
+const (
+	VUnknown    VerClass = iota
+	V1                   = iota
+	V2PreV264            = iota
+	V2PostV264           = iota
+	V3orGreater          = iota
+)
+
+// Check the passed interfaces to determine the Version to decide if any
+// migration is needed
+func getClusterVersion(ctx context.Context, c client.Interface, cv1 ClientV1) (VerClass, string) {
```

Comment:
Would a better interface for this function be something like `requiresMigration` that just returns a boolean?

I feel like this function is trying to do too much. We could have nice composable parts:

```go
func getClusterVersion(ctx context.Context, client ClientV1) (string, error) {...}
func requiresMigration(v string) bool {...}
```